File: frontend/src/api/stock.ts
import client from './client'

export interface KlineItem {
  trade_date: string
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface StockScore {
  code: string
  stock_name: string
  score: number
  tech_score: number
  fund_score: number
  momentum_score: number
  sentiment_score: number
  reason: string
}

export interface StockSignal {
  trade_date: string
  direction: string
  score: number
  reason: string
}

export interface StockSearchResult {
  code: string
  name: string
  market: string
  industry: string
}

export function searchStocks(q: string) {
  return client.get<StockSearchResult[]>('/stock/search', { params: { q } })
}

export function fetchKline(code: string, days = 60) {
  return client.get<KlineItem[]>(`/stock/${code}/kline`, { params: { days } })
}

export function fetchScore(code: string) {
  return client.get<StockScore>(`/stock/${code}/score`)
}

export function fetchSignals(code: string) {
  return client.get<StockSignal[]>(`/stock/${code}/signals`)
}

export interface CompareStock {
  code: string
  name: string
  industry: string
  score: number
  tech_score: number
  fund_score: number
  momentum_score: number
  sentiment_score: number
  klines: { trade_date: string; close: number; volume: number; change_pct: number }[]
}

export function fetchCompare(codes: string[]) {
  return client.get<CompareStock[]>('/stock/compare', { params: { codes: codes.join(',') } })
}

File: backend/app/api/stock.py
from datetime import date, timedelta

from fastapi import APIRouter, Depends, Query
from sqlalchemy import or_
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.models.signal import Signal
from app.models.stock import StockBasic, StockDaily
from app.models.system import User

router = APIRouter(prefix="/api/stock", tags=["stock"])


@router.get("/search")
def search(
    q: str = Query(..., min_length=1, max_length=20),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    pattern = f"%{q}%"
    results = (
        db.query(StockBasic)
        .filter(or_(StockBasic.code.like(pattern), StockBasic.name.like(pattern)))
        .limit(20)
        .all()
    )
    return [
        {
            "code": s.code,
            "name": s.name,
            "market": s.market,
            "industry": s.industry or "",
        }
        for s in results
    ]


@router.get("/{code}/kline")
def kline(
    code: str,
    days: int = Query(60, ge=1, le=250),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    cutoff = date.today() - timedelta(days=int(days * 1.5))
    items = (
        db.query(StockDaily)
        .filter(StockDaily.code == code, StockDaily.trade_date >= cutoff)
        .order_by(StockDaily.trade_date)
        .all()
    )
    return [
        {
            "trade_date": str(k.trade_date),
            "open": k.open,
            "high": k.high,
            "low": k.low,
            "close": k.close,
            "volume": k.volume,
        }
        for k in items
    ]


@router.get("/{code}/signals")
def signals(
    code: str,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    items = (
        db.query(Signal)
        .filter(Signal.code == code)
        .order_by(Signal.trade_date.desc())
        .limit(30)
        .all()
    )
    return [
        {
            "trade_date": str(s.trade_date),
            "direction": s.direction,
            "score": s.score,
            "reason": s.reason,
        }
        for s in items
    ]


@router.get("/{code}/score")
def score(
    code: str,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    s = db.query(Signal).filter(Signal.code == code).order_by(Signal.trade_date.desc()).first()
    if not s:
        return {"code": code, "score": 0, "detail": "无评分数据"}
    return {
        "code": s.code,
        "stock_name": s.stock_name,
        "score": s.score,
        "tech_score": s.tech_score,
        "fund_score": s.fund_score,
        "momentum_score": s.momentum_score,
        "sentiment_score": s.sentiment_score,
        "reason": s.reason,
    }
